crates/ruma-events/src/secret/request.rs

```diff
@@ -0,0 +1,381 @@
+                    },
+                };
+
+                let requesting_device_id: DeviceIdBox = match requesting_device_id {
+                    None => {
+                        return Err(de::Error::missing_field("requesting_device_id"));
+                    }
+                    Some(id) => id.into(),
+                };
+
+                let request_id = match request_id {
+                    None => {
+                        return Err(de::Error::missing_field("request_id"));
+                    }
+                    Some(id) => id,
+                };
+
+                Ok(RequestToDeviceEventContent::new(
+                    request_action,
+                    requesting_device_id,
+                    request_id,
+                ))
+            }
+        }
+
+        const FIELDS: &[&str] = &["name", "action", "requesting_device_id", "request_id"];
+        deserializer.deserialize_struct(
+            "RequestToDeviceEventContent",
+            FIELDS,
+            RequestToDeviceEventContentVisitor,
+        )
+    }
+}
+
+/// Action for a *m.secret.request* event.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum RequestAction {
+    /// Request a secret by its name.
+    Request(SecretName),
+
+    /// Cancel a request for a secret.
+    RequestCancellation,
+
+    #[doc(hidden)]
+    _Custom(String),
+}
+
+impl Serialize for RequestAction {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            Self::Request(name) => {
+                let mut st = serializer.serialize_map(Some(2))?;
+                st.serialize_entry("name", name)?;
+                st.serialize_entry("action", "request")?;
+                st.end()
+            }
+            Self::RequestCancellation => {
+                let mut st = serializer.serialize_map(Some(1))?;
+                st.serialize_entry("action", "request_cancellation")?;
+                st.end()
+            }
+            RequestAction::_Custom(custom) => serializer.serialize_str(custom),
+        }
+    }
+}
+
+/// The name of a secret.
+#[derive(Clone, Debug, StringEnum)]
+pub enum SecretName {
+    /// Cross-signing master key (m.cross_signing.master).
+    #[ruma_enum(rename = "m.cross_signing.master")]
+    CrossSigningMasterKey,
+
+    /// Cross-signing user-signing key (m.cross_signing.user_signing).
+    #[ruma_enum(rename = "m.cross_signing.user_signing")]
+    CrossSigningUserSigningKey,
+
+    /// Cross-signing self-signing key (m.cross_signing.self_signing).
+    #[ruma_enum(rename = "m.cross_signing.self_signing")]
+    CrossSigningSelfSigningKey,
+
+    /// Recovery key (m.megolm_backup.v1).
+    #[ruma_enum(rename = "m.megolm_backup.v1")]
+    RecoveryKey,
+
+    /// Custom secret name.
+    Custom(String),
```

Comment:
Like in the other similar enum types, the `Custom` variant should be `#[doc(hidden)]` and prefixed with an underscore to make it very clear it's not meant to be used.

Reply:
The `SecretName` variants aren't mentioned in the "Secrets" module of the unstable spec, but rather in the "End-to-End Encryption" module. That's why I thought that these were only suggestions, and it was expected that any other secret could be stored.
However I see that hiding the `_Custom` variant does not change the functionality, and simplifies the docs for those who only want to implement the "basic" features.

Reply:
Regardless of whether they're suggestions or not, we must not expose `Custom` variants except for some special cases where we know no additional variants are going to ever be introduced, for forwards compatibility. Having `"foo"` deserialize to `SomeEnum::Custom("foo")` first and to `SomeEnum::Foo` later after an update means people's code checking for `SomeEnum::Custom("foo")` will break, without there being any compiler warnings, which is very annoying.